src/charmed_analytics_ci/git_client.py

```diff
@@ -0,0 +1,291 @@
+
+    Attributes:
+        username: GitHub username.
+        token: Personal access token used for authentication.
+    """
+
+    username: str
+    token: str
+
+
+class GitClientError(Exception):
+    """
+    Exception raised for generic Git client-related issues.
+
+    Raised when an operation with git or GitHub API fails due to configuration or connectivity.
+    """
+
+
+class PullRequestAlreadyExistsError(GitClientError):
+    """
+    Raised when a pull request from the current branch already exists on GitHub.
+
+    Attributes:
+        url: URL of the existing pull request.
+    """
+
+    def __init__(self, url: str):
+        super().__init__(f"A pull request already exists: {url}")
+        self.url = url
+
+
+class GitClient:
+    """
+    A client that wraps local Git operations and GitHub pull request creation.
+
+    Attributes:
+        repo: A GitPython Repo object representing the local repository.
+        gh_repo: A PyGithub Repository object connected to the GitHub repo.
+        credentials: GitHub credentials used for Git and API operations.
+    """
+
+    def __init__(self, repo: Repo, gh_repo: Repository, credentials: GitCredentials) -> None:
+        """
+        Initialize a GitClient instance.
+
+        Args:
+            repo: The local GitPython repository instance.
+            gh_repo: The corresponding GitHub Repository object.
+            credentials: GitHub credentials for authentication.
+        """
+        self.repo = repo
+        self.gh_repo = gh_repo
+        self.credentials = credentials
+
+    @property
+    def current_branch(self) -> str:
+        """
+        Return the name of the currently checked out Git branch.
+        """
+        return self.repo.active_branch.name
+
+    def checkout_branch(self, branch: str) -> None:
+        """
+        Switch to a specified branch, creating it locally if it doesn't exist.
+
+        Args:
+            branch: Name of the branch to check out.
+
+        Raises:
+            GitClientError: If checkout fails for unexpected reasons.
+        """
+        try:
+            self.repo.git.checkout(branch)
+        except GitCommandError as e:
+            # Check if the error was due to a missing branch
+            error_msg = str(e).lower()
+            if "did not match any file(s) known to git" in error_msg or "pathspec" in error_msg:
+                logger.info(f"Branch '{branch}' not found; creating new local branch.")
+                try:
+                    self.repo.git.checkout("-b", branch)
+                except GitCommandError as create_err:
+                    raise GitClientError(
+                        f"Failed to create new branch '{branch}': {create_err}"
+                    ) from create_err
+            else:
+                raise GitClientError(f"Failed to checkout branch '{branch}': {e}") from e
+
+    def commit_and_push(
+        self, commit_message: str, branch: Optional[str] = None, force: bool = False
```

Comment:
question: is there a use for `force` here? since we do not create a PR if one already exists. I'm a little cautious with having force push in an automated tool.